### PR TITLE
[Issue #59] Adding a bunch of BiMaps

### PR DESCRIPTION
**Pull request by jiegillet**
Resovles #59 

Pretty straightforward.
- I'm not 100% sure about was where to add the `bytestring` dependency in the `.cabal` files. Please double check.
- I'm not including the List, Set and HashSet discussed in #59 yet.
- This will have conflicts with #106, but I'll fix whatever needed when/if needed
- I haven't commented the functions. Should I go with the standard `'XXX' bimap for 'AnyValue'. Usually used with 'arrayOf' combinator.` everywhere?

### CODE REVIEW

**Comment by jiegillet** [comment]
I've modified the PR to adjust to #95.
Should I add `_List` `_Set` and `_HashSet` as discussed in #59? They are probably what makes most sense in the current context: give me the conversion bimap for en element and I'll give you a list.

File: src/Toml/BiMap.hs [hunk 1]
```diff
@@ -145,6 +157,31 @@ _ShowString = BiMap (Just . show) readMaybe
...
+          | otherwise = Just (toInteger n)
+
+_Natural :: BiMap Natural AnyValue
+_Natural = _UnsignedInteger >>> _Integer
+
+_Word :: BiMap Word AnyValue
+_Word = _UnsignedInteger >>> _Integer
+
+_RealReal ::  (Real a, Real b, Fractional a, Fractional b) => BiMap a b
+_RealReal = iso realToFrac realToFrac
```
Comment (ChShersh) [inline src/Toml/BiMap.hs]: I feel like `_RealReal` is not going to be used outside `_Float`...

**Comment by ChShersh** [comment]
> Should I add _List _Set and _HashSet as discussed in #59?

This is a hard question. As it was discussed, it doesn't really makes sense to use `Read/Show` for custom data types like this one:

```haskell
data User = User
    { userFirstName :: Text
    , userLastName :: Text
    , userFollowers :: Natural
    }
```

But we still can have sets via `Array` for primitive types, but I propose to somehow reflect this via name. Like `_ArraySet` and `_ArrayHashSet`.

**Comment by jiegillet** [comment]
These type of custom values are much more like a `TOML` table than an `Anything`.

**Comment by ChShersh** [comment]
> These type of custom values are much more like a TOML table than an Anything.

Yes, and currently `tomland` doesn't allow to parse values of type `[User]`. But TOML format supports this with the Array of Tables (especially inline tables).

But Array of Tables is not yet implemented in the `tomland`:

* https://github.com/kowainik/tomland/issues/13

**Comment by jiegillet** [comment]
what do you think of this function:
```haskell
_Pair :: BiMap a AnyValue -> BiMap b AnyValue -> BiMap (a, b) AnyValue
_Pair (BiMap av va) (BiMap bv vb) = BiMap for back
  where
  for (a, b) = case (av a, bv b) of
    (Just (AnyValue v1), Just (AnyValue v2)) ->
      Just $ AnyValue $ Array [Array [v1], Array [v2]]
    _ -> Nothing
  back (AnyValue v) = case v of
    Array [Array [v1], Array [v2]] ->
      (,) <$> va (AnyValue v1) <*> vb (AnyValue v2)
    _ -> Nothing
```
It would be a step forward maybe. If we could generalize it to more than 2 values it could be neat. 
Note: the pattern matching for `Array [...]` in `back` requires the extension `GADTs`.

File: src/Toml/BiMap.hs [hunk 2]
```diff
@@ -145,6 +157,31 @@ _ShowString = BiMap (Just . show) readMaybe
...
+
+_Float :: BiMap Float AnyValue
+_Float = _RealReal >>> _Double
+
+_ByteStringText :: BiMap ByteString Text
+_ByteStringText = prism encodeUtf8 maybeText
+  where maybeText = either (const Nothing) Just . decodeUtf8'
+
+_ByteString:: BiMap ByteString AnyValue
+_ByteString = _ByteStringText >>> _Text
```
[the remark this answers is not in the document]
Comment (jiegillet) [reply src/Toml/BiMap.hs]: Done!

File: src/Toml/BiMap.hs (hunk 1 above)
Comment (jiegillet) [reply src/Toml/BiMap.hs]: That's very likely. I removed it from the list and only use it in `_Float`.

File: src/Toml/BiMap.hs [hunk 3]
```diff
@@ -157,31 +161,42 @@ _ShowString = BiMap (Just . show) readMaybe
 _Show :: (Show a, Read a) => BiMap a AnyValue
 _Show = _ShowString >>> _String
 
-_UnsignedInteger :: (Integral n) => BiMap n Integer
-_UnsignedInteger = prism fromIntegral maybeInteger
+_NaturalInteger :: BiMap Natural Integer
+_NaturalInteger = BiMap (Just . toInteger) maybeInteger
   where maybeInteger n
```
Comment (ChShersh) [inline src/Toml/BiMap.hs]: I would appreciate having explicit type signature for `maybeInteger` function to make it clear in what direction numbers are moving. It's quite easy to swap arguments of `_NaturalInteger` bimap and not notice it... So something like this really improves code readability:

```haskell
  where
    maybeInteger :: Integer -> Maybe Natural
    maybeInteger n
         | n < 0 ....
```

**Comment by ChShersh** [comment]
@jiegillet I was thinking on using array of arrays to encode lists of custom data types. But this becomes too complicated. And it loses important property of TOML: human-readability.

It's actually quite easy to implement list combinator but it requires support for array of tables (#13). Currently you're thinking on the `BiMap` level and we already have `_Array`:

```haskell
_Array :: BiMap a AnyValue -> BiMap [a] AnyValue
```

`BiMap` works on the level of `AnyValue` and this level is for primitive types like `Integer`, `Double`, `Text`, `DateTime`, `Bool`. To parse custom user data types, we need to use `TomlCodec` from `Toml.Bi.Combinators` library. So if we have user data type like this one:

```haskell
data User = User
    { userFirstName :: Text
    , userLastName :: Text
    , userFollowers :: Natural
    }
```

we currently can write bidirectional TOML converter for this data type in the following way:

```haskell
userCodec :: TomlCodec User
userCodec = User
    <$> Toml.text "firstName" .= userFirstName
    <*> Toml.text "lastName" .= userLastName
    <*> Toml.natural "followers" .= userFollowers
```

And then this data type can be parsed from the following TOML:

```toml
firstName = "Dmitrii"
lastName  = "Kovanikov"
followers = 500
```

But this TOML file represents only single `User` data type. Though, it's possible to represent list of users in the following way in TOML using _Array of tables_:

```toml
[[user]]
firstName = "Dmitrii"
lastName  = "Kovanikov"
followers = 500

[[user]]
firstName = "Foo"
lastName  = "Bar"
followers = 256
```

This is a list in TOML that contains two users. It's can be represented even better with _inline tables_:

```toml
users =
    [ { firstName = "Dmitrii", lastName = "Kovanikov", followers = 500 }
    , { firstName = "Foo", lastName = "Bar", followers = 256 }
    ]
```

But `tomland` doesn't support currently neither array of tables nor inline tables. Though, there're open issues :slightly_smiling_face:

**Comment by jiegillet** [comment]
You're right that working at the `BiMap` levels gets too complicated.

I still haven't added `_Set` and `_HashSet` but they're ready to go if you want them.

I was thinking of working on parsing inlines tables and/or array of tables. I know someone already volunteered but they haven't been active. I'll see.

**Comment by ChShersh** [comment]
> I still haven't added _Set and _HashSet but they're ready to go if you want them.

I will create separate issue for that 👍
I think this can be done separately.

> I was thinking of working on parsing inlines tables and/or array of tables. I know someone already volunteered but they haven't been active. I'll see.

The issue was blocked because of #99. But now #116 is merged, so you can try to ask about array of tables again 🙂

**Pull request merged**